dotnet/src/SemanticKernel/Reliability/DefaultHttpRetryPolicy.cs

```diff
@@ -0,0 +1,185 @@
+    /// If the exception contains a RetryAfter header, the action is retried after the specified delay.
+    /// If configured to use exponential backoff, the delay is doubled for each retry.
+    /// If the action throws a critical exception, it is not retried.
+    /// </remark>
+    /// <param name="request">The request to execute</param>
+    /// <param name="log">The logger to use</param>
+    /// <param name="cancellationToken">The cancellation token</param>
+    public async Task<HttpResponseMessage> ExecuteWithRetryAsync(Func<Task<HttpResponseMessage>> request, ILogger log,
+        CancellationToken cancellationToken = default)
+    {
+        int retryCount = 0;
+
+        var start = this._timeProvider.GetCurrentTime();
+        while (true)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            TimeSpan waitFor = default;
+            string reason = string.Empty;
+            HttpResponseMessage? response = null;
+            try
+            {
+                response = await request();
+
+                // If the request does not require a retry then we're done
+                if (!this.ShouldRetry(response.StatusCode))
+                {
+                    return response;
+                }
+
+                // Drain response content to free connections. Need to perform this
+                // before retry attempt and before the TooManyRetries ServiceException.
+                if (response.Content != null)
+                {
+#if NET5_0_OR_GREATER
+                        await response.Content.ReadAsByteArrayAsync(cancellationToken).ConfigureAwait(false);
+#else
+                    await response.Content.ReadAsByteArrayAsync().ConfigureAwait(false);
+#endif
+                }
+
+                reason = response.StatusCode.ToString();
+
+                if (retryCount >= this._config.MaxRetryCount)
+                {
+                    log.LogWarning(
+                        "Error executing request, max retry count reached. Reason: {0}", reason);
+                    return response;
+                }
+
+                // If the retry delay is longer than the total timeout, then we'll
+                // just return
+                if (!this.HasTimeForRetry(start, retryCount, response, out waitFor))
+                {
+                    log.LogWarning(
+                        "Error executing request, max total retry time reached. Reason: {0}", reason);
+                    return response;
+                }
+            }
+            catch (Exception e) when ((this.ShouldRetry(e) || this.ShouldRetry(e.InnerException)) &&
+                                      retryCount < this._config.MaxRetryCount &&
+                                      this.HasTimeForRetry(start, retryCount, response, out waitFor))
+            {
+                reason = e.GetType().ToString();
+            }
+
+            // If the request requires a retry then we'll retry
```

Comment:
Right now I'm not cloning the request -- would like to better understand why we should and then add the implementation.

Reply:
Hmm I might have got myself stuck here by making things wrap the PostAsync method and not a delegatinghandler. May go ahead and disregard that for now depending on what folks think.